File: java/com/dragon/inter/Logical.java
package com.dragon.inter;

import com.dragon.lexer.Token;
import com.dragon.symbols.Type;

public class Logical extends Expr {
  protected Expr left;
  protected Expr right;

  public Logical(Token tok, Expr xl, Expr xr) {
    super(tok, null);
    left = xl;
    right = xr;
    type = check(left.type, right.type);
    if (type == null) {
      error("Type Error");
    }
  }

  public Type check(Type tLeft, Type tRight) {
    if (tLeft == Type.Bool && tRight == Type.Bool) {
      return Type.Bool;
    }
    return null;
  }

  @Override
  public Expr gen(StringBuilder b) {
    var f = newLabel();
    var a = newLabel();
    var tmp = new Temp(type);
    jumping(b, 0, f);
    emit(b, tmp.toString() + " = true");
    emit(b, "goto L" + a);
    emitLabel(b, f);
    emit(b, tmp.toString() + " = false");
    emitLabel(b, a);
    return tmp;
  }

  @Override
  public String toString() {
    return left.toString() + " " + op.toString() + " " + right.toString();
  }
}

File: cpp/inter/logical.cpp
#include "cpp/inter/logical.hpp"

#include "cpp/inter/temporary.hpp"

namespace inter {
std::shared_ptr<Expression> Logical::gen(std::stringstream &ss) {
  auto f = new_label();
  auto a = new_label();
  auto temp = Temporary::create(type());
  jumping(ss, 0, f);
  emit(ss, temp->to_string() + " = true");
  std::stringstream lbl;
  lbl << a;
  emit(ss, "goto L" + lbl.str());
  emit_label(ss, f);
  emit(ss, temp->to_string() + " = false");
  emit_label(ss, a);
  return temp;
}
}  // namespace inter
